### PR TITLE
Add pre-commit workflow & fix remaining pyright issues

### PR DESCRIPTION
**Pull request by younik**
Aș per title, this PR adds the pre-commit workflow & fixes the remaining typing issues

### CODE REVIEW

File: .github/workflows/pre-commit.yml [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+  contents: read
+
+jobs:
+  pre-commit:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+      - uses: actions/setup-python@v4
+        with:
+          python-version: '3.11'
```
Comment (josephdviviano) [inline .github/workflows/pre-commit.yml:17]: unlikely to be an issue but the codebase requires >=3.10 so maybe we should use that here.

File: src/gfn/containers/replay_buffer.py [hunk 2]
```diff
@@ -180,6 +180,12 @@ def add(self, training_objects: Transitions | Trajectories | tuple[States]):
 
         # Our buffer is full and we will prioritize diverse, high reward additions.
         else:
+            if (
+                self.training_objects.log_rewards is None
+                or training_objects.log_rewards is None
+            ):
+                raise ValueError("log_rewards must be defined for prioritized replay.")
```
Comment (josephdviviano) [inline src/gfn/containers/replay_buffer.py:187]: good catch thank you

**Review by josephdviviano — CHANGES_REQUESTED** [review]
LGTM! Just one comment re python versions. Thank you!

File: .github/workflows/pre-commit.yml (hunk 1 above)
Comment (younik) [reply .github/workflows/pre-commit.yml]: Changed to use Python 3.10

**Review by josephdviviano — APPROVED** [review]
LGTM!

**Pull request merged**